doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+``git@`` addresses, so that git knows to connect via SSH instead of HTTPS. For
+example:
+
+.. code-block:: console
 
-Style
-^^^^^
-* Use single quotes whenever possible.
-* Prefer generator or list comprehensions over ``filter``, ``map`` and other functional idioms.
-* Use explicit functional constructors for builtin containers to improve readability (e.g., ``list()``, ``dict``).
-* Avoid nested functions or class methods if possible -- use private functions instead.
-* Avoid ``**kwargs`` and ``*args`` in function signatures.
-* Add brief docstrings to simple private functions and complete docstrings for complex ones.
+    $ git remote -v  # show existing remote addresses
+    $ git remote set-url origin git@github.com:$GITHUB_USERNAME/mne-python.git
+    $ git remote set-url upstream git@github.com:mne-tools/mne-python.git
 
-Checking and building documentation
------------------------------------
 
-All changes to the codebase must be properly documented.
-To ensure that documentation is rendered correctly, the best bet is to
-follow the existing examples for class and function docstrings,
-and examples and tutorials.
+MNE-Python coding conventions
+=============================
 
-Our documentation (including docstring in code) uses ReStructuredText format,
-see `Sphinx documentation`_ to learn more about editing them. Our code
-follows the `NumPy docstring standard`_.
+General requirements
+^^^^^^^^^^^^^^^^^^^^
 
-Documentation is automatically built remotely during pull requests. If
-you want to also test documentation locally, you will need to install (using
-``conda`` or ``pip`` as preferred):
+All new functionality must have test coverage
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For example, a new :class:`mne.Evoked` method in :file:`mne/evoked.py` should
+have a corresponding test in :file:`mne/tests/test_evoked.py`.
+
+
+All new functionality must be documented
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This includes thorough docstring descriptions for all public API changes, as
+well as how-to examples or longer tutorials for major contributions. Docstrings
+for private functions may be more sparse, but should not be omitted.
+
+
+New API elements should be added to the master reference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Classes, functions, methods, and attributes cannot be cross-referenced unless
+they are included in the :doc:`python_reference`
+(:file:`doc/python_reference.rst`).
+
+
+Describe your changes in the changelog
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Include in your changeset a brief description of the change in the
+:doc:`changelog <whats_new>` (:file:`doc/whats_new.rst`; this can be skipped
+for very minor changes like correcting typos in the documentation). Note that
+there are sections of the changelog for each release, and separate subsections
+for bugfixes, new features, and changes to the public API. It is usually best
```

Comment:
I was going to say we should add:

... public API:

- Changelog
    For new features.
- Bug
    For bug fixes to existing features.
- API
    For backward-incompatible changes.

But we could just put these as RST comments in the `whats_new.rst` itself, so this doc does not grow.